src/v/cloud_topics/frontend/frontend.cc

```diff
@@ -311,10 +315,148 @@ frontend::make_l1_reader(cloud_topic_log_reader_config& cfg) const {
+    // we need to first query L0, then we can query L1. In this case, the data
+    // intervals that we queried between L0 and L1 could overlap, but that's OK
+    // because we just favor using L1 to answer the time query in this case.
+    auto l0_result = co_await l0_timequery(cfg);
+    auto l1_result = co_await l1_timequery(cfg);
+    vlog(
+      cd_log.trace,
+      "metadata timequery for L1: {}, for L0: {}",
+      l1_result,
+      l0_result);
+    if (l1_result) {
+        co_return co_await refine_timequery_result(
+          *l1_result, cfg.abort_source);
+    }
+    if (l0_result) {
+        co_return co_await refine_timequery_result(
+          *l0_result, cfg.abort_source);
+    }
+    co_return std::nullopt;
+}
+
+ss::future<std::optional<frontend::coarse_grained_timequery_result>>
+frontend::l1_timequery(storage::timequery_config cfg) {
+    auto ct_state = _partition->get_cloud_topics_state();
+    auto l1_metastore = ct_state->local().get_l1_metastore();
+    auto maybe_tidp = ntp_to_topic_id_partition(_partition->ntp());
+    vassert(
+      maybe_tidp.has_value(),
+      "No topic id for cloud topic {}",
+      _partition->ntp());
+    const auto& tidp = *maybe_tidp;
+    // I don't love this, but we clamp min/max offsets by the kafka start offset
+    // and the LSO/HWM, but we can ignore the max offset for L1 because we never
+    // upload anything less than LSO to L1.
+    std::ignore = cfg.max_offset;
+    // Go query our start offset from the L1 metastore
+    auto result = co_await l1_metastore->get_first_ge(
+      tidp, model::offset_cast(cfg.min_offset), cfg.time);
+    if (!result.has_value()) {
+        if (
+          result.error() == l1::metastore::errc::out_of_range
+          || result.error() == l1::metastore::errc::missing_ntp) {
+            co_return std::nullopt;
+        }
+        throw std::runtime_error(fmt_with_ctx(
+          fmt::format,
+          "unable to read from l1 to service timequery @ {} for: {}",
+          cfg.time,
+          _partition->ntp()));
+    }
+    co_return coarse_grained_timequery_result{
+      .time = cfg.time,
+      .start_offset = result->first_offset,
+      .last_offset = result->last_offset,
+    };
+}
+
+ss::future<std::optional<frontend::coarse_grained_timequery_result>>
+frontend::l0_timequery(storage::timequery_config cfg) {
+    // Read L0 metadata to find the right batch. We can't use
+    // _partition->timequery because it will filter for only data batches, not
+    // placeholder batches.
+    auto reader = co_await _partition->make_local_reader({
+      /*start_offset=*/cfg.min_offset,
+      /*max_offset=*/cfg.max_offset,
+      /*max_bytes=*/std::numeric_limits<size_t>::max(),
+      /*type_filter=*/std::nullopt,
+      /*time=*/cfg.time,
+      /*as=*/cfg.abort_source,
+      /*client_addr=*/cfg.client_address,
+    });
+    auto type_filter = std::to_array({
+      model::record_batch_type::raft_data,
```

Comment:
Just checking my understanding -- is the raft_data filter here meant for transactional control batches? Or are those placeholders as well?

Reply:
it's for transactional control batches correct. I should probably test that